src/lsp/cobol_lsp/lsp_request.ml

```diff
@@ -586,6 +586,82 @@ let handle_document_symbol registry (params: DocumentSymbolParams.t) =
       Some (`DocumentSymbol symbols)
     end
 
+(** { Document Code Lens } *)
+
+module PosSet = Set.Make(struct
+    type t = Position.t
+    let compare (p1: t) (p2: t) =
+      let c = p2.line - p1.line in
+      if c <> 0 then c else p2.character - p1.character
+  end)
+
+let codelens_positions ~uri group =
+  let filename = Lsp.Uri.to_path uri in
+  let open struct
+    include Cobol_common.Visitor
+    include Cobol_data.Visitor
+    type context =
+      | ProcedureDiv
+      | DataDiv
+      | None
+  end in
+  let set_context context (old, acc) =
+    do_children_and_then (context, acc) (fun (_, acc) -> (old, acc))
+  in
+  let take_when_in context { loc; _ } (current, acc) =
+    if context <> current
+    then skip (current, acc)
+    else
+      let range = Lsp_position.range_of_srcloc_in ~filename loc in
+      skip (context, PosSet.add range.start acc)
+  in
+  Cobol_unit.Visitor.fold_unit_group
+    object (v)
+      inherit [_] Cobol_unit.Visitor.folder
+      method! fold_procedure _ = set_context ProcedureDiv
+      method! fold_data_definitions _ = set_context DataDiv
+      method! fold_paragraph' _ = skip
+      method! fold_procedure_name' = take_when_in ProcedureDiv
+      method! fold_qualname' = take_when_in DataDiv
+      method! fold_record_renaming { renaming_name; _ } =
+        take_when_in DataDiv renaming_name
+      method! fold_field_definition' { payload = field; _ } acc =
+        fold_field_definition v { field with field_redefines = None } acc
+        |> skip
+      method! fold_table_definition' { payload = table; _ } acc =
+        fold_table_definition v { table with table_redefines = None } acc
+        |> skip
```

Comment:
Even if that leads to longer code, I think it would be clearer to explicitly fold over selected fields of `field` (resp. `table`). I think "only" `field_qualname`, `field_layout`, `field_conditions`, and `field_redefintions`. Then with an explicit:
```ocaml
ignore (field_redefines, field_leading_ranges,
        field_offset, field_size);
```
that'll help in case we re-define parts of the type.